### PR TITLE
PURRSOFT 79 - Implement entity foster animal mapping

### PR DESCRIPTION
**Pull request by Domynis**
- added new AnimalFosterMap entity for MtoM relation
- added Controller, CommandHandler and Queries

### CODE REVIEW

File: PurrSoft/PurrSoft.Domain/Entities/AnimalFosterMap.cs [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+
+public class AnimalFosterMap
+{
+	public Guid Id { get; set; }
+	public DateTime StartFosteringDate { get; set; }
+	public DateTime? EndFosteringDate { get; set; }
+	public string? SupervisingComment { get; set; }
+
+	public Guid AnimalId { get; set; }
+	public Animal Animal { get; set; }
```
Comment (raulfrandes) [inline PurrSoft/PurrSoft.Domain/Entities/AnimalFosterMap.cs:11]: Objects should be virtual

File: PurrSoft/PurrSoft.Domain/Entities/AnimalFosterMap.cs [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+
+	public string FosterId { get; set; }
+	public Foster Foster { get; set; }
```
Comment (raulfrandes) [inline PurrSoft/PurrSoft.Domain/Entities/AnimalFosterMap.cs:14]: Here too

File: PurrSoft/PurrSoft.Domain/Entities/Animal.cs [hunk 3]
```diff
@@ -4,11 +4,16 @@ namespace PurrSoft.Domain.Entities;
...
-    public Boolean Sterilized { get; set; }
-    public string? ImageUrl { get; set; }
+	public Guid Id { get; set; }
+	public AnimalType AnimalType { get; set; }
+	public string? Name { get; set; }
+	public int YearOfBirth { get; set; }
+	public string? Gender { get; set; }
+	public Boolean Sterilized { get; set; }
+	public string? ImageUrl { get; set; }
+	public IList<AnimalFosterMap> FosteredBy { get; set; }
```
Comment (raulfrandes) [inline PurrSoft/PurrSoft.Domain/Entities/Animal.cs:14]: List objects should be virtual

File: PurrSoft/PurrSoft.Application/Queries/AnimalFosterMapQueries/AnimalFosterMapQueries.cs [hunk 4]
```diff
@@ -0,0 +1,33 @@
+﻿using PurrSoft.Application.Common;
+using PurrSoft.Application.Models;
+using PurrSoft.Application.QueryOverviews;
+
+namespace PurrSoft.Application.Queries.AnimalFosterMapQueries;
+
```
Comment (raulfrandes) [inline PurrSoft/PurrSoft.Application/Queries/AnimalFosterMapQueries/AnimalFosterMapQueries.cs:6]: Are the first two getters necessary if we have GetFilteredAnimalFosterMapsQueries with the filter fields AnimalId and FosterId?

File: PurrSoft/PurrSoft.Api/Controllers/AnimalFosterMapController.cs [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+
+namespace PurrSoft.Api.Controllers;
+
+[Route("api/[controller]")]
+[ApiController]
+public class AnimalFosterMapController : BaseController
+{
+	[HttpGet()]
+	[Authorize(AuthenticationSchemes = "Bearer", Roles = "Manager, Foster")]
+	[ProducesResponseType(typeof(CollectionResponse<AnimalFosterMapDto>), (int)HttpStatusCode.OK)]
```
Comment (raulfrandes) [inline PurrSoft/PurrSoft.Api/Controllers/AnimalFosterMapController.cs:20]: Your response type does not match your return type, it says that it should response with a CollectionResponse of AnimalFosterMapDto, but it returns a CollectionResponse of AnimalFosterMapOverview.

File: PurrSoft/PurrSoft.Api/Controllers/AnimalFosterMapController.cs [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+	public async Task<CollectionResponse<AnimalFosterMapOverview>> GetAnimalFosterMaps([FromQuery] GetFilteredAnimalFosterMapsQueries query)
+	{
+		return await Mediator.Send(query, new CancellationToken());
+	}
+
+	[HttpGet("{id}")]
+	[Authorize(AuthenticationSchemes = "Bearer", Roles = "Manager, Foster")]
+	[ProducesResponseType(typeof(AnimalFosterMapDto), (int)HttpStatusCode.OK)]
+	[ProducesResponseType((int)HttpStatusCode.BadRequest)]
```
Comment (raulfrandes) [inline PurrSoft/PurrSoft.Api/Controllers/AnimalFosterMapController.cs:29]: You could say that the type of the response is a CommandResponse

File: PurrSoft/PurrSoft.Api/Controllers/AnimalFosterMapController.cs [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+		catch (ValidationException ex)
+		{
+			return BadRequest(new CommandResponse(ex.Errors.ToList()));
+		}
+	}
+
+	[HttpPost()]
+	[Authorize(AuthenticationSchemes = "Bearer", Roles = "Manager, Foster")]
+	[ProducesResponseType(typeof(CommandResponse), (int)HttpStatusCode.OK)]
+	[ProducesResponseType((int)HttpStatusCode.BadRequest)]
```
Comment (raulfrandes) [inline PurrSoft/PurrSoft.Api/Controllers/AnimalFosterMapController.cs:53]: Here too

**Review by raulfrandes — CHANGES_REQUESTED** [review]

File: PurrSoft/PurrSoft.Application/Queries/AnimalFosterMapQueries/AnimalFosterMapQueries.cs (hunk 4 above)
Comment (Domynis) [reply PurrSoft/PurrSoft.Application/Queries/AnimalFosterMapQueries/AnimalFosterMapQueries.cs:6]: No, I initially used them but then forgot to delete them.

File: PurrSoft/PurrSoft.Api/Controllers/AnimalFosterMapController.cs (hunk 6 above)
Comment (Domynis) [reply PurrSoft/PurrSoft.Api/Controllers/AnimalFosterMapController.cs:29]: Isn't it fine for a typeof(AnimalFosterMapDto) response if that is the entity returned with the Ok Status code?

File: PurrSoft/PurrSoft.Api/Controllers/AnimalFosterMapController.cs (hunk 7 above)
Comment (Domynis) [reply PurrSoft/PurrSoft.Api/Controllers/AnimalFosterMapController.cs:53]: Is something like this fine?

`[HttpPost()]
[Authorize(AuthenticationSchemes = "Bearer", Roles = "Manager, Foster")]
[ProducesResponseType(typeof(CommandResponse), (int)HttpStatusCode.OK)]
[ProducesResponseType(typeof(CommandResponse), (int)HttpStatusCode.BadRequest)]`

File: PurrSoft/PurrSoft.Api/Controllers/AnimalFosterMapController.cs (hunk 6 above)
Comment (raulfrandes) [reply PurrSoft/PurrSoft.Api/Controllers/AnimalFosterMapController.cs:29]: i was referring to the BadRequest response, sorry for not specifying

**Comment by florinacho337** [comment]
When you put an invalid foster Id, POST endpoint returns a 500 error:
Screenshot from 2024-12-08 12-04-08

**Comment by Domynis** [comment]
> When you put an invalid foster Id, POST endpoint returns a 500 error: Screenshot from 2024-12-08 12-04-08

The errors that relate to EF always return 500 at the moment. We should treat all such cases in a different task / PR. As such, the scope of this PR relates only to the AnimalFosterMap entity.

**Review by florinacho337 — APPROVED** [review]

**Pull request merged**